base_layer/core/src/transactions/transaction_protocol/sender.rs

```diff
@@ -774,4 +782,86 @@ mod test {
+            .with_fee_per_gram(fee_per_gram)
+            .with_offset(a.offset.clone())
+            .with_private_nonce(a.nonce.clone())
+            .with_change_secret(a.change_key.clone())
+            .with_input(utxo.clone(), input)
+            .with_amount(0, amount);
+        match builder.build::<Blake256>(&factories) {
+            Ok(_) => panic!("'BuildError(\"Fee is greater than amount\")' not caught"),
+            Err(e) => assert_eq!(e.message, "Fee is greater than amount".to_string()),
+        };
+    }
+
+    #[test]
+    fn allow_fee_larger_than_amount() {
+        let factories = CryptoFactories::default();
+        // Alice's parameters
+        let a = TestParams::new();
+        // Bob's parameters
+        let b = TestParams::new();
+        let (utxo_amount, fee_per_gram, amount) = get_fee_larger_than_amount_values();
+        let (utxo, input) = make_input(&mut OsRng, utxo_amount, &factories.commitment);
+        let mut builder = SenderTransactionProtocol::builder(1);
+        builder
+            .with_lock_height(0)
+            .with_fee_per_gram(fee_per_gram)
+            .with_offset(a.offset.clone())
+            .with_private_nonce(a.nonce.clone())
+            .with_change_secret(a.change_key.clone())
+            .with_input(utxo.clone(), input)
+            .with_amount(0, amount)
+            .with_prevent_fee_gt_amount(false);
+        let mut alice = match builder.build::<Blake256>(&factories) {
+            Ok(s) => s,
+            Err(e) => panic!("Unexpected error: {:?}", e),
+        };
+        assert!(alice.is_single_round_message_ready());
+        let msg = alice.build_single_round_message().unwrap();
+
+        // Send message down the wire....and wait for response
+        assert!(alice.is_collecting_single_signature());
+
+        // Test serializing the current state to be sent and resuming from that serialized data
+        let ser = alice.save_pending_transaction_to_be_sent().unwrap();
+        let mut alice = SenderTransactionProtocol::load_pending_transaction_to_be_sent(ser).unwrap();
+
+        // Receiver gets message, deserializes it etc, and creates his response
+        let bob_info = SingleReceiverTransactionProtocol::create(
+            &msg,
+            b.nonce,
+            b.spend_key,
+            OutputFeatures::default(),
+            &factories,
+        )
+        .unwrap();
+        // Alice gets message back, deserializes it, etc
+        alice
+            .add_single_recipient_info(bob_info.clone(), &factories.range_proof)
+            .unwrap();
+        // Transaction should be complete
+        assert!(alice.is_finalizing());
+        // Fee larger than amount should be allowed
+        assert!(alice.get_fee_amount().unwrap() > utxo_amount - alice.get_fee_amount().unwrap());
+        match alice.finalize(KernelFeatures::empty(), &factories) {
+            Ok(_) => (),
+            Err(e) => panic!("Unexpected error: {:?}", e),
+        };
+    }
```

Comment:
There is a specific feature to test in `fn validate(&self)` that is called by `pub fn finalize(&mut self, features: KernelFeatures, factories: &CryptoFactories)`, so that is why I need to test that the code does not panic.
``` rust
   fn validate(&self) -> Result<(), TPE> {
        if let SenderState::Finalizing(info) = &self.state {
            let total_amount = calculate_total_amount(info.amount_to_self, info.amounts.clone());
            let fee = info.metadata.fee;
            // The fee should be less than the amount. This isn't a protocol requirement, but it's what you want 99.999%
            // of the time, however, always preventing this will also prevent spending dust in some edge cases.
            if fee > total_amount {
                warn!(
                    target: LOG_TARGET,
                    "Fee ({}) is greater than amount ({}) for Transaction (TxId: {}).", fee, total_amount, info.ids[0]
                );
                if info.prevent_fee_gt_amount {
                    return Err(TPE::ValidationError("Fee is greater than amount".into()));
                }
            }
```

Reply:
Removed latter part of the test in line with change in code